Feat/tanlang/add more cmd / 增加更多功能和命令

LinZexiao opened the pull request:
## Linked Issue

ref https://github.com/filecoin-project/venus/issues/5304


## Detail
### 新增命令
- 消息的查询
- replace 消息
- 发消息地址管理
- 查看订单
- 升级订单
- 给矿工设置 ask
- 创建矿工
- 查看矿工证明dealline
- 查看扇区详情
- 扇区续期

LinZexiao committed:
feat: add cmd to query and replace msg

LinZexiao committed:
feat: add cmd to manage address

LinZexiao committed:
feat: add cmd to mange ask of miner

LinZexiao committed:
feat: add cmd to manage deal

LinZexiao committed:
feat: add cmd to create miner

LinZexiao committed:
feat: add cmd to extend sector

LinZexiao committed:
feat: add cmd to query deadline and sector
of miner

LinZexiao committed:
feat: make msg list more practical

LinZexiao committed:
fix: explain wrapped func correctly

LinZexiao committed:
fix: recover version test before gen api client

simlecode commented on cmd/cli/deal.go:
```diff
@@ -0,0 +1,189 @@
+					"%s\t%d\t%s\t%s\t%s\t%d\t%d\t%d\t%s\n",
+					deal.Receiver.String(),
+					deal.ID,
+					"..."+payloadCid[len(payloadCid)-8:],
```
这里 verbose 不起效

LinZexiao replied:
retrieval 订单信息比较少,所以 verbose  flag 只对 storage 的订单生效

simlecode replied:
我的意思是用户会不会需要完整的payloadCid的值

LinZexiao replied:
已修复

simlecode commented on cmd/cli/miner.go:
```diff
@@ -0,0 +1,449 @@
+	},
+}
+
+var minerSetAskCmd = &cli.Command{
```
考虑分开来设置，flag 夹杂在一起

diwufeiwen commented on client/client.go:
```diff
@@ -2,35 +2,153 @@ package client
+
 type Client struct {
 	*resty.Client
+	apiVersion string
```
要不要给默认值，假如忘记调用SetVersion呢

LinZexiao replied:
空值作为默认值?  因为就客户端本身来说,版本前缀的形式不是固定的,也可以是没有版本前缀的

diwufeiwen commented on client/client.go:
```diff
@@ -2,35 +2,153 @@ package client
+	c.apiVersion = version
+}
+
+func (c *Client) Post(ctx context.Context, path string, body, result interface{}) error {
```
这里有个问题，作为调用者，怎么知道每个请求的body, result传什么结构的数据？

diwufeiwen replied:
或者说，是不是缺少类型的验证，如果调用者乱传呢

LinZexiao replied:
入参和返回结果,应该 参考 服务层的接口, 入参的校验也会放在 服务层的接口来做.
传统的前后端应用 应该生成接口文档,可以考虑在后续的开发中补充这部分的内容

diwufeiwen commented on cmd/cli/addr.go:
```diff
@@ -0,0 +1,241 @@
+var addrListCmd = &cli.Command{
+	Name:      "list",
+	Usage:     "list address",
+	ArgsUsage: "[address]",
```
这个是说请求一个指定地址的addr的是先获取所有的然后再帅选？是不是作为参数传进去，接口内部实现比较好

LinZexiao replied:
这个命令设计之初就是希望列出所有地址的, 加了个 address 可以让用户进行筛选是为了让用户使用更加便利,所以从思路上,这个做法还算恰当? 
另外,考虑到一个用户其实发消息的地址不多,所以过滤这个操作是否在接口内部实现,影响不大

diwufeiwen replied:
这样没啥问题

diwufeiwen commented on cmd/cli/addr.go:
```diff
@@ -0,0 +1,241 @@
+			return err
+		}
+
+		return nil
```
给用户一个成功的反馈

LinZexiao replied:
已修改

diwufeiwen commented on cmd/cli/addr.go:
```diff
@@ -0,0 +1,241 @@
+			params["SelectMsgNum"] = ctx.Uint64("num")
+		}
+
+		if err != nil {
```
err的判断跟调用断了？

LinZexiao replied:
已更正

diwufeiwen commented on cmd/cli/addr.go:
```diff
@@ -0,0 +1,241 @@
+			"Address": addr,
+		}
+
+		isSetSpec := ctx.IsSet("gas-overestimation") || ctx.IsSet("gas-feecap") || ctx.IsSet("max-fee") || ctx.IsSet("base-fee") || ctx.IsSet("gas-over-premium")
```
IsSetSpec这个来判断是不是设置了gas相关参数？能从变量本身判断不

LinZexiao replied:
因为这几个参数没有明显的 "空值" ,所以不好根据入参判断是否设置过该字段

diwufeiwen commented on cmd/cli/addr.go:
```diff
@@ -0,0 +1,241 @@
+			return err
+		}
+
+		return client.Post(ctx.Context, "/addr/operate", params, nil)
```
这里貌似存在所有flag都未设置的情况，这种调用了会怎样？

LinZexiao replied:
服务端不会做任何操作,不过这种空操作请求,应该被避免掉
已修改

diwufeiwen commented on cmd/cli/cli.go:
```diff
@@ -1,21 +1,278 @@
+	msgTypes "github.com/filecoin-project/venus/venus-shared/types/messager"
 )
 
+var log = logging.Logger("cmd")
```
cli？

LinZexiao replied:
已修改

diwufeiwen commented on cmd/cli/cli.go:
```diff
@@ -1,21 +1,278 @@
+
+		_, _ = fmt.Fprintf(w, "%s\t%d\t%s\t%s\t%s\t%s\t%s\t%s\t%s", propcid, deal.DealID, storagemarket.DealStates[deal.State], deal.PieceStatus,
+			deal.Proposal.Client, deal.Proposal.Provider, units.BytesSize(float64(deal.Proposal.PieceSize)), fil, deal.Proposal.Duration())
+		if verbose {
```
以下内容是不是没有verbose

LinZexiao replied:
下面补充了 几个字段的内容,也就是 verbose 的内容

diwufeiwen commented on cmd/cli/deal.go:
```diff
@@ -0,0 +1,189 @@
+var dealListCmd = &cli.Command{
+	Name:      "list",
+	Usage:     "List deals",
+	ArgsUsage: "[miner address]",
```
这里应该是可以多个，但是下面逻辑又只取第一个，不合逻辑。要不这里只能是一个<miner address> 要么就允许差多个矿工的？

LinZexiao replied:
已更正

diwufeiwen commented on cmd/cli/deal.go:
```diff
@@ -0,0 +1,189 @@
+			return nil
+		}
+
+		return client.Post(cctx.Context, "/deal/storage/state", &service.StorageDealUpdateStateReq{
```
应该给一个显示的成功的反馈

LinZexiao replied:
已修复

diwufeiwen commented on cmd/cli/deal.go:
```diff
@@ -0,0 +1,189 @@
+	},
+}
+
+var dealUpdateCmd = &cli.Command{
```
这个命令是不是应该带miner，做权限验证，只允许修改自己矿工的deal状态？或者已经通过token判断了

LinZexiao replied:
正常的使用流程中, 用户只能通过 vt 拿到自己的 订单id, 所以正常使用流程是不会出现修改不属于自己订单的情况的.
如果说 用户通过别的方式获取到他人订单id,想要修改,这层防护是由 market 的接口层来完成的

diwufeiwen commented on cmd/cli/miner.go:
```diff
@@ -0,0 +1,449 @@
+var minerDeadlineCmd = &cli.Command{
+	Name:      "deadline",
+	Usage:     "query miner proving deadline info",
+	ArgsUsage: "[minerAddress]",
```
<minerAddress>

LinZexiao replied:
这里的 description 是指 cli.Command 用来描述参数的 Description 吗,如果是的话 miner address 这样的名字自描述性应该够了?

diwufeiwen commented on cmd/cli/sector.go:
```diff
@@ -0,0 +1,119 @@
+			return fmt.Errorf("failed to print json: %w", err)
+		}
+
+		return nil
```
显示的成功反馈

LinZexiao replied:
成功反馈: printJSON(resp)
失败反馈: return err

diwufeiwen commented on repo/config/config.go:
```diff
@@ -6,19 +6,22 @@ import (
 	NodeAPI     *APIInfo
 	MessagerAPI *APIInfo
 	MarketAPI   *APIInfo
+	Wallets     []address.Address
```
这两个配置的用途是啥？他们是不是存在一定的关联？比如wa 归属于 ma

LinZexiao replied:
这两个配置的作用是让用户可以提供一组自己的矿工地址和钱包地址,在某些情况下,可以被 vt 用作默认的矿工地址和钱包地址.
钱包地址主要是用户用于签名发消息的地址, 和矿工地址之间可能存在关联,但不是必然的

hunjixin replied:
听上去是个想当然的需求

LinZexiao replied:
还是有需要的,包括到后面开发到第二阶段的时候,也还是需要存储用户的这些数据的

diwufeiwen replied:
这么说可以先留着，后面再完善

diwufeiwen commented on route/route.go:
```diff
@@ -21,8 +21,39 @@ func registerRoute(s *service.Service) http.Handler {
+	chainGroup.GET("/head/", Wrap(s.ChainHead))
+
+	msgGroup := apiV0Group.Group("/msg")
+	msgGroup.GET("query", Wrap(s.MsgQuery))
```
query -> /query

LinZexiao replied:
gin 对于子路由项 加不加斜杠都是兼容的, 我的想法是,对于 Group 这种路由聚合的子项, 加斜杆.  不是路由聚合的子项则不加斜杆, 用于区分和理解,这样可以吗?

diwufeiwen reviewed (commented):
首先，这个开发速度我觉得很强，非常利害！
建议：
1. 一个pr不要提如此多，可以分模块命令提
2. 提交pr的是第一责任人，希望这个pr的命令首先自己测试过的

hunjixin commented on route/wrap.go:
```diff
@@ -77,13 +81,31 @@ func Wrap(fn interface{}) gin.HandlerFunc {
+
+			var err error
+
+			if ctx.Request.ContentLength > 0 {
```
这样写感觉还是蛮奇怪的，这意思就是客户端有几种传值的办法？

LinZexiao replied:
是的,尽可能兼容 多种传参方式,因为不能直接判断 是不是带了 body,就只能用间接的方法了

hunjixin commented on service/service.go:
```diff
@@ -36,3 +69,332 @@ func (s *Service) Send(ctx context.Context, params *SendParams) (string, error)
+	return cid, err
+}
+
+func (s *Service) AddrOperate(ctx context.Context, params *AddrsOperateReq) error {
```
待商榷的功能，目前只需要给个更新参数就行，其他的没必要做。

LinZexiao replied:
目前已经实现了的话,不确定要不要就先保留,明确不要再删掉?

hunjixin commented on service/service.go:
```diff
@@ -36,3 +69,332 @@ func (s *Service) Send(ctx context.Context, params *SendParams) (string, error)
+		}
+	}
+
+	ret, err := s.Messager.GetMessageByUid(ctx, params.MsgId)
```
不用等待？

LinZexiao replied:
不需要等待,这个接口是无状态可重入的, 只要客户端那边 定时轮询就好

hunjixin commented on route/route.go:
```diff
@@ -21,8 +21,39 @@ func registerRoute(s *service.Service) http.Handler {
+
+	dealGroup := apiV0Group.Group("/deal")
+	dealGroup.GET("storage", Wrap(s.DealStorageList))
+	dealGroup.GET("retrieval", Wrap(s.DealRetrievalList))
```
检索的接口目前应该没法判断检索订单到底是谁的。

LinZexiao replied:
是的,所以目前返回的结果是所有的,这个要去掉吗?

LinZexiao committed:
fix: apply review suggestions
supply feedback after cmd exec
change logger name to cli from cmd
fix verbose flag
fix err check

diwufeiwen commented on cmd/cli/addr.go:
```diff
@@ -0,0 +1,251 @@
+		}
+
+		if ctx.Args().Len() > 0 {
+			addr, err := address.NewFromString(ctx.Args().First())
```
对于参数的判断提前，如果错误，不请求接口

LinZexiao replied:
已修改

diwufeiwen commented on cmd/cli/miner.go:
```diff
@@ -0,0 +1,462 @@
+			err = client.Post(ctx, "/miner/create", map[string]string{"MsgId": params.MsgId}, &miner)
+		}
+
+		fmt.Println(miner)
```
如果 err != nil && !strings.Contains(err.Error(), "temp error") 时代表创建矿工失败，输出错误信息？

LinZexiao replied:
目前来说,临时错误意味着创建矿工的消息还没上链,需要等待,所以在这个时候只需要在 log 里面记录一下.然后等待一个时间再查询一下就好了

diwufeiwen replied:
除去临时错误外，不一定就创建成功了，还有一类就是因某种原因失败了

diwufeiwen replied:
就是for循环的退出不一定代表miner创建成功，还有可能是失败了

diwufeiwen replied:
就是for循环的退出不一定代表miner创建成功，还有可能是失败了

LinZexiao replied:
已修正

LinZexiao committed:
fix: report err when create miner fail

diwufeiwen reviewed (approved):
LGTM

LinZexiao committed:
feat: rm timeout from config

LinZexiao committed:
feat: add tool to gen IServiceStruct

LinZexiao committed:
feat: auto register route

LinZexiao committed:
feat: replace client with api

LinZexiao committed:
feat: apply review suggestion
return err rather than panic when create client

LinZexiao committed:
fix: rm blank line

LinZexiao committed:
feat: replace function name toMap with toParamsMap

LinZexiao committed:
Merge pull request #3 from ipfs-force-community/feat/tanlang/replace-client-with-api

Feat/tanlang/replace client with api / 用封装好的 api 代替 httpClient
